src/libmeasurement_kit/neubot/negotiate_impl.hpp

```diff
@@ -0,0 +1,213 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include "src/libmeasurement_kit/common/utils.hpp"
+#include <functional>
+#include <iostream>
+#include <measurement_kit/common.hpp>
+#include <measurement_kit/http.hpp>
+#include <measurement_kit/mlabns.hpp>
+#include <measurement_kit/neubot.hpp>
+#include <measurement_kit/report.hpp>
+#include <stdlib.h>
+#include <string>
+#include <unistd.h>
+
+#define DASH_MAX_NEGOTIATION 512
+
+using namespace mk;
+using namespace mk::neubot;
+using namespace mk::net;
+using namespace mk::http;
+using namespace mk::report;
+using namespace mk::mlabns;
+
+namespace mk {
+namespace neubot {
+namespace negotiate {
+
+template <MK_MOCK_NAMESPACE(http, request_sendrecv)>
+void collect(Var<Transport> transport, Callback<Error> cb, std::string auth,
+             Var<Entry> measurements, Settings settings, Var<Reactor> reactor,
+             Var<Logger> logger) {
+    std::string body = measurements->dump();
+    settings["http/method"] = "POST";
+    settings["http/path"] = "/collect/dash";
+
+    request_sendrecv(
+        transport, settings,
+        {
+            {"Content-Type", "application/json"}, {"Authorization", auth},
+        },
+        body,
+        [=](Error error, Var<Response> res) {
+            if (error || (res->status_code != 200)) {
```

Comment:
Note to self:

- `res` in theory *may* be `nullptr`
- we don't close the connection in this code path